src/invariant/BucketListIsConsistentWithDatabase.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "bucket/Bucket.h"
+#include "database/Database.h"
+#include "invariant/BucketListIsConsistentWithDatabase.h"
+#include "invariant/InvariantManager.h"
+#include "ledger/AccountFrame.h"
+#include "ledger/DataFrame.h"
+#include "ledger/OfferFrame.h"
+#include "ledger/TrustFrame.h"
+#include "lib/util/format.h"
+#include "main/Application.h"
+
+namespace stellar
+{
+
+const std::string BucketListIsConsistentWithDatabase::kName =
+    "BucketListIsConsistentWithDatabase";
+
+std::shared_ptr<Invariant>
+BucketListIsConsistentWithDatabase::registerInvariant(Application& app)
+{
+    return app.getInvariantManager()
+              .registerInvariant<BucketListIsConsistentWithDatabase>(
+                        app.getDatabase());
+}
+
+BucketListIsConsistentWithDatabase::BucketListIsConsistentWithDatabase(
+    Database& db)
+    : mDb{db}
+{
+}
+
+std::string
+BucketListIsConsistentWithDatabase::getName() const
+{
+    return kName;
+}
+
+std::string
+BucketListIsConsistentWithDatabase::checkOnBucketApply(
+        std::shared_ptr<Bucket> bucket)
+{
+    Bucket::InputIterator iter(bucket);
+    if (!Bucket::InputIterator(bucket))
+    {
+        return {};
+    }
+
+    uint64_t nAccounts = 0, nTrustLines = 0, nOffers = 0, nData = 0;
+    uint32 modifiedSinceLedgerSeq = std::numeric_limits<uint32>::max();
```

Comment:
from what I can tell the logic around `modifiedSinceLedgerSeq` is not correct, the value of `modifiedSinceLedgerSeq` is defined by which bucket we're checking in the [context of a bucketlist](https://github.com/stellar/stellar-core/blob/master/src/bucket/BucketList.h#L106-L108)
With the expected range figured out, you can then:
* ensure that the entries in this bucket are in the expected range (this probably belongs in the regular apply code, not in an invariant)
* query the SQL database for entries that fall into the range (well, with the upper bound set to max int this code)